python/pyspark/sql/connect/sql_formatter.py

```diff
@@ -0,0 +1,82 @@
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import string
+import typing
+from typing import Any, Optional, List, Tuple, Sequence, Mapping
+import uuid
+
+from pyspark.sql.connect.proto import Expression
+
+if typing.TYPE_CHECKING:
+    from pyspark.sql.connect import SparkSession, DataFrame
+from pyspark.sql.connect.functions import lit
+
+
+class SQLStringFormatter(string.Formatter):
+    """
+    A standard ``string.Formatter`` in Python that can understand PySpark instances
+    with basic Python objects. This object has to be clear after the use for single SQL
+    query; cannot be reused across multiple SQL queries without cleaning.
+    """
+
+    def __init__(self, session: "SparkSession") -> None:
+        self._session: "SparkSession" = session
+        self._temp_views: List[Tuple[DataFrame, str]] = []
+        self._unresolvedColumns: typing.Dict[str, Expression] = {}
+
+    def get_field(self, field_name: str, args: Sequence[Any], kwargs: Mapping[str, Any]) -> Any:
+        obj, first = super(SQLStringFormatter, self).get_field(field_name, args, kwargs)
+        return self._convert_value(obj, field_name), first
+
+    def _convert_value(self, val: Any, field_name: str) -> Optional[str]:
+        """
+        Converts the given value into a SQL string.
+        """
+        from pyspark.sql.connect import Column, DataFrame
+
+        if isinstance(val, DataFrame):
+            for df, n in self._temp_views:
+                if df is val:
+                    return n
+            df_name = "_pyspark_%s" % str(uuid.uuid4()).replace("-", "")
+            self._temp_views.append((val, df_name))
+            val.createOrReplaceTempView(df_name)
+            return df_name
+        if isinstance(val, Column):
+            for f, _ in self._unresolvedColumns.items():
+                if f is field_name:
+                    break
+            self._unresolvedColumns[field_name] = val.expr.to_plan(self._session.client)
+            return wrap(field_name)
+        elif isinstance(val, str):
+            self._unresolvedColumns[field_name] = lit(val).expr.to_plan(self._session.client)
+            return wrap(field_name)
+        return val
+
+    def clear(self) -> None:
+        """TODO We need to find a better time to drop view"""
```

Comment:
There is something wrong with this version of the code
1. There is no way to drop view normally. Because the view is to pass the name into the sql for reference, not the logical plan of the reference, if the view is dropped before the action, the view cannot be processed normally. The difference between this and the original pyspark program is that the original pyspark saves the parsed logical plan, but now we save the name of the view. Connect has no way to save the parsed logical plan. If it does so, it will increase the complexity of the client side, because the original logical plan and the protobuf type plan must be maintained at the same time.
2. In order to solve this problem, `ParameterizedQuery` needs to be extended to support adding logical plans as parameters, just like the current `args`

@HyukjinKwon @cloud-fan @MaxGekk WDYT?

Reply:
Hm, can't use just drop by the view by name? IIRC, Spark Connect's `sql` method itself eagerly executes the commands so the views should be created with `clear` is invoked.

Reply:
Eagerly executes not suit for SELECT query. If the sql are SELECT statement, clear view will cause VIEW_NOT_FOUND

Reply:
I think you can work around by calling `df.schema()` that triggers analysis for now, and commenting that this is a workaround to be fixed.

Reply:
Sorry, I don't get the point. Where to call `df.schema()`? And the calling will resolved what problem? Like this?
<img width="401" alt="image" src="https://user-images.githubusercontent.com/32387433/231373441-75bb89a8-2c90-4ad0-9a4b-0cf60421dd86.png">

Reply:
Not work, can't pass Analysis stage when use action.

Reply:
Update: I try to extended `ParameterizedQuery` to add dataframes, but have one blocker is that parser does not recognize `{}` symbols. Extend parser will be a big change for other module. Can I remove support for dataframes in connect client sql temporary? @HyukjinKwon